Add health checking infrastructure to SPIRE Server

mcpherrinm opened the pull request:
We add a new common health package which is a SPIRE-flavored wrapper around
github.com/InVisionApp/go-health

This doesn't actually check anything yet.  I started looking at how to have something that called a Status method on plugins if they exposed it, but I got hung up in the weeds figuring out how the catalog worked.  So I thought it was better to start small and get some early feedback if this approach seems reasonable.

evan2645 commented on pkg/common/health/health.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func (c *Checker) AddCheck(name string, checker health.ICheckable) error {
+	return c.hc.AddCheck(&health.Config{
```
Quick review of go-health source leads me to believe that AddCheck isn't thread safe. Should we have a mutex on Checker struct?

mcpherrinm replied:
AddCheck can only be called before healthchecks are started, IE, it must be called during server startup.  I think we're single threaded there, but it's certainly safer to throw a mutex around it.

evan2645 commented on pkg/common/health/health.go:
```diff
@@ -0,0 +1,95 @@
+		handler := http.NewServeMux()
+
+		handler.HandleFunc(config.getHealthPath(), handlers.NewJSONHandlerFunc(hc, nil))
+		handler.HandleFunc(config.getLivePath(), hello)
```
Based on this, I assume that Health == k8s readiness and Live == k8s liveness?

Perhaps we should converge on the k8s naming for these two checks to avoid confusion... same for default paths e.g. /ready ?

mcpherrinm replied:
Yes, I'm happy to use /ready and /live (and corresponding configurable names) to be consistent with k8s.

mcpherrinm replied:
Done, code uses ready and live now.

evan2645 commented on cmd/spire-server/cli/run/run.go:
```diff
@@ -34,6 +35,7 @@ type runConfig struct {
 	Server        serverRunConfig            `hcl:"server"`
 	PluginConfigs catalog.HCLPluginConfigMap `hcl:"plugins"`
 	Telemetry     telemetry.FileConfig       `hcl:"telemetry"`
+	HealthChecks  health.Config              `hcl:"health_checks"`
```
I am curious to hear your thoughts on how this might apply to spire agent in the future. I'm personally a little hesitant to allow the agent to bind to a non-localhost address.

If we think that there might be differences in what is made available to the server vs the agent, then this config might better suited to be put directly into the `server { ... }` and `agent { ... }` config blocks.

Lastly (as a general comment) we have been trying to keep configuration complexity to a minimum. The more that we can have sane defaults for, the better. The less we can expose that has the potential for misconfiguration or abuse, the better.

evan2645 replied:
Here's one possible example:
```
server {
    healthcheck_listener_enabled = true
    healthcheck_bind_address = "0.0.0.0:80"
}

...

agent {
    healthcheck_listener_enabled = true
}

mcpherrinm replied:
My use-case is fine with plaintext being localhost-only, and this will match the prometheus one (just set a port).

mcpherrinm replied:
Current config sample:

```health_checks {
       listener_enabled = true
       bind_address = "127.0.0.1"
       bind_port = "80"
       ready_path = "/ready"
       live_path = "/live"
}```

It's server-only, but we could disallow setting bind_address in the agent if you want (though similarly to the discussion in prometheus metrics, not sure that's worthwhile

evan2645 commented on pkg/common/health/config.go:
```diff
@@ -0,0 +1,39 @@
+import "time"
+
+type Config struct {
+	Address  string        `hcl:"listen_address"`
```
Should we have a localhost default here? In exchange for using listen address we could add an `enabled` bool

mcpherrinm replied:
My use-case is fine with plaintext being localhost-only, and this will match the prometheus one (just set a port).

mcpherrinm replied:
this is now configurable as discussed above

evan2645 commented on pkg/common/health/config.go:
```diff
@@ -0,0 +1,39 @@
+
+type Config struct {
+	Address  string        `hcl:"listen_address"`
+	Interval time.Duration `hcl:"interval"`
```
Any thoughts on what circumstances this might be tweaked? I wouldn't mind a hardcoded value here, a user setting this to 1s or less seems like it could be bad

side-thought: any idea what hcl does for time.Duration type? This old issue seems to suggest that duration is expressed as nanoseconds (!)
https://github.com/hashicorp/hcl/issues/202

mcpherrinm replied:
Hardcoding a value seems OK to me too, though I think if we do that, the individual checks should hardcode them (Eg, if somebody pays per request to an upstream CA, you may want to health check every hour instead of every 5 seconds).

evan2645 replied:
Hmm yes that's a good point. I don't think that should be too problematic. A plugin implementation with extra constraints like that could also decouple internally

mcpherrinm replied:
Each check now specifies its own interval (though no checks are added yet, except a base dummy one)

evan2645 commented on pkg/common/health/health.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func (c *Checker) ListenAndServe(ctx context.Context) error {
+	if err := c.hc.Start(); err != nil {
```
Enabling health check listener should be optional.

Given util.RunTasks consideration, perhaps we don't bother setting up any of the health.New() stuff in NewChecker in the event that we are disabled, and AddCheck and ListenAndServe return nil?

mcpherrinm replied:
If no address is specified, it doesn't listen -- it'll still log healthy/unhealthy state transitions though, which seemed valuable to me.  I don't have strong opinions on what the level of configiness should be here.

evan2645 replied:
Aah gotcha sorry I missed the conditional below. Yes logging health state changes seems valuable regardless of whether the listener is enabled or not - thanks for pointing that out.

mcpherrinm replied:
It's explicitly configurable now

evan2645 commented on pkg/common/health/health.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
```
I think this statement should be one line below, inside the conditional, so we don't add a waiter if c.server is nil.

evan2645 commented on pkg/common/health/health.go:
```diff
@@ -0,0 +1,95 @@
+		defer wg.Done()
+		<-ctx.Done()
+		if c.server != nil {
+			c.server.Shutdown(ctx)
```
Context will already be cancelled at this point - perhaps we should call Close instead?

mcpherrinm replied:
done

evan2645 commented:
Quick ping on this @mcpherrinm - did my comments look ok to you and are you planning on pulling this up? It's been open for a while, just want to make sure you're not blocked on us. We can try to pick it up at some point but are a little backlogged at the moment.

mcpherrinm commented:
I'm planning to pick this up; we're running with this as a patch for the moment but want to get something merged here.  I'm just busy with all the things we need to do for getting SPIRE running right now.

evan2645 commented:
> I'm planning to pick this up

Ok great, no worries just wanted to check in.

> I'm just busy with all the things we need to do for getting SPIRE running right now.

Anything in particular we can help out with? Any pains that could be lessened?

mcpherrinm commented:
^ most recent change is just a rebase onto current master before I pick this up again

mcpherrinm commented:
I've addressed most of the feedback I think.  the fixup commits don't have DCOs; but I was planning to squash them.  Should I squash them now, or would it be useful for reviewing to keep them separate temporarily?

mcpherrinm commented:
I've rebased again & addressed all the feedback I think.

ajessup commented:
This is awesome to see. Thanks @mcpherrinm! 

Am I correct in understanding this isn't yet user-facing functionality? (ie. there are no docs to update?)

mcpherrinm commented:
Yup, this doesn't really do much yet.  I just want to start laying some groundwork.

Subsequent changes can start actually adding some checks, and then we can document how to turn on health checks once they actually do something.

ajessup commented:
SGTM

On Fri, Jul 19, 2019, 2:41 PM Matthew McPherrin <notifications@github.com>
wrote:

> Yup, this doesn't really do much yet. I just want to start laying some
> groundwork.
>
> Subsequent changes can start actually adding some checks, and then we can
> document how to turn on health checks once they actually do something.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/spiffe/spire/pull/882?email_source=notifications&email_token=AACHGYZQYNKZMBRUIISW2L3QAIYITA5CNFSM4HJJUGKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2M2WEY#issuecomment-513387283>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACHGY2334ICHPITIOTLUK3QAIYITANCNFSM4HJJUGKA>
> .
>

mcpherrinm commented:
@evan2645 Any more feedback here?

azdagron commented on pkg/common/health/logger.go:
```diff
@@ -0,0 +1,33 @@
+
+// HealthCheckFailed is triggered when a health check fails the first time
+func (sl *statusListener) HealthCheckFailed(entry *health.State) {
+	log.With("check", entry.Name).
```
sl.log...

azdagron commented on pkg/common/health/logger.go:
```diff
@@ -0,0 +1,33 @@
+
+// HealthCheckRecovered is triggered when a health check recovers
+func (sl *statusListener) HealthCheckRecovered(entry *health.State, recordedFailures int64, failureDurationSeconds float64) {
+	log.With("check", entry.Name).
```
sl.log...

mcpherrinm replied:
Fixed these two.  This was also dragging in an extra dep on promethus/common/log, intead of using logrus directly.

mcpherrinm commented:
It looks like the build failed in the MySQL integration test, because it wasn't able to connect to a mysql server.  I think that's probably unrelated?

evan2645 commented:
I restarted the build and recorded the output. I think I've seen it once or
twice before... I'll file an issue tomorrow.

--
evan

On Mon, Jul 22, 2019, 17:04 Matthew McPherrin <notifications@github.com>
wrote:

> It looks like the build failed in the MySQL integration test, because it
> wasn't able to connect to a mysql server. I think that's probably unrelated?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spiffe/spire/pull/882?email_source=notifications&email_token=AAO2UTUKLVAE65VMOJTBGXDQAZDITA5CNFSM4HJJUGKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2RQIGA#issuecomment-513999896>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO2UTWCFSB5GV75CFZX5K3QAZDITANCNFSM4HJJUGKA>
> .
>
